.circleci/config.yml

```diff
@@ -1,48 +1,130 @@
 version: 2.1
 
+executors:
+  python:
+    parameters:
+      version:
+        type: string
+        default: "3.7"
+
+    docker:
+      - image: circleci/python:<< parameters.version >>
+
+commands:
+  build_package:
+    steps:
+      - run: |
+          python setup.py sdist
+          PACKAGE=$(python setup.py --fullname)
+          tar -zxvf "dist/${PACKAGE}.tar.gz"
+          mv $PACKAGE package
+
+  installation:
+    parameters:
+      test_requirements:
+        type: boolean
+        default: false
+
+    steps:
+      - checkout
+      - build_package
+      - run:
+          name: "Create virtual environment and install requirements."
+          command: |
+            virtualenv env -q
+            source env/bin/activate
+            pip config --site set global.progress_bar off
+            pip install package/
+
+      - when:
+          condition: << parameters.test_requirements >>
+          steps:
+            - run:
+                name: "Install test requirements."
+                command: |
+                  source env/bin/activate
+                  pip install -r package/dev-requirements.txt
+
 jobs:
-  test_python:
+  lint_tests:
+    parameters:
+      python_version:
+        type: string
+        default: "3.7"
+
+    executor:
+      name: python
+      version: << parameters.python_version >>
+
+    steps:
+      - installation:
+          test_requirements: true
+
+      - run:
+          name: "Run lint tests."
+          command: |
+            source env/bin/activate
+            make lint-tests
+
+  unit_tests:
     parameters:
       python_version:
         type: string
         default: "3.7"
+
       codecov:
         type: boolean
         default: false
-    working_directory: ~/compose
-    docker:
-        - image: circleci/python:<< parameters.python_version >>
+
+    executor:
+      name: python
+      version: << parameters.python_version >>
+
     steps:
-      - checkout
-      - run: |
-          virtualenv test_python -q
-          source test_python/bin/activate
-          make installdeps
-      - run: |
-          source test_python/bin/activate
-          make lint
-      - run: |
-          source test_python/bin/activate
-          coverage erase
-          make test
+      - installation:
+          test_requirements: true
+
       - when:
+          name: "Run unit tests with code coverage."
           condition: << parameters.codecov >>
           steps:
             - run: |
-                source test_python/bin/activate
+                source env/bin/activate
+                make unit-tests ADDOPTS="--cov=composeml"
```

Comment:
the other step to ensure we test the packaged version is to run the tests in the `package` folder instead of the root folder, so that we are using the packages tests and avoid using the local version of the composeml code

Reply:
Should the packaged version contain the make file?

Reply:
After a local run, it seems like the package doesn't contain the make file, but we can add in the manifest.

Reply:
I believe you can specify the path to the make file, so that we don't need to package it

Reply:
Or there's the `make -C`  option which tells make to change directories before doing anything

Reply:
Yes, that's right we can specify the path to the make file. I think the `make -C` will still run at the root level -- it only changes directory before running. I think this command should work:

```
make unit-tests -f ../Makefile 
```